### ISSUE TITLE
[BUG] Unable to upgrade Endpoint via Agent

### ISSUE THREAD

**Issue by ferullo** [elastic/beats#24477]
I am unable to upgrade Endpoint via Agent from 7.11.2-SNAPSHOT to 7.12.0-SNAPSHOT.

Steps to reproduce
* Deploy a 7.12.0-SNAPSHOT Stack
* Enroll a 7.11.2-SNAPSHOT Agent on Windows 10 x64 with test signing enabled
* Add Endpoint Security Integration
* See Agent and Endpoint deploy successfully. `elastic-agent.exe version` shows 7.11.2-SNAPSHOT. `elastic-endpoint.exe version` shows 7.11.2-SNAPSHOT
* Initiate an update to 7.12.0-SNAPSHOT via Fleet
* See Agent upgrade to 7.12.0-SNAPSHOT. `elastic-agent.exe version` shows 7.11.2-SNAPSHOT.
* (BUG) Endpoint remains on 7.11.2-SNAPSHOT. `elastic-endpoint.exe version` shows 7.11.2-SNAPSHOT.
* (BUG) Agent enters a HEALTHY state in Fleet despite not having upgraded Endpoint.
* (BUG) Agent's version in Fleet shows 7.12.0 even though not all components have been upgraded

Relevant logs from Agent are below. I do not see any reason to think Endpoint failed to upgrade or interfered with Agent's upgrade.

```
2021-03-09T14:05:31.145-0800    INFO    operation/operator.go:259       operation 'operation-install' skipped for metricbeat.7.12.0-SNAPSHOT
2021-03-09T14:05:31.145-0800    INFO    operation/operator.go:259       operation 'operation-start' skipped for metricbeat.7.12.0-SNAPSHOT
2021-03-09T14:05:31.146-0800    INFO    operation/operation_fetch.go:61 endpoint-security.7.12.0-SNAPSHOT already exists in C:\Program Files\Elastic\Agent\data\elastic-agent-af32fe\downloads\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64.zip. Skipping operation operation-fetch
2021-03-09T14:05:31.146-0800    INFO    operation/operation_fetch.go:61 endpoint-security.7.12.0-SNAPSHOT already exists in C:\Program Files\Elastic\Agent\data\elastic-agent-af32fe\downloads\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64.zip. Skipping operation operation-fetch
2021-03-09T14:05:32.916-0800    INFO    log/reporter.go:40      2021-03-09T14:05:32-08:00: type: 'STATE': sub_type: 'RUNNING' message: Application: metricbeat--7.12.0-SNAPSHOT[15e46460-8123-11eb-b9c4-e16b4b3269c3]: State changed to RUNNING: Running
2021-03-09T14:05:33.106-0800    ERROR   application/emitter.go:107      Failed to render configuration with latest context from composable controller: operator: failed to execute step sc-run, error: rename C:\Program Files\Elastic\Agent\data\elastic-agent-af32fe\install\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64 C:\Program Files\Elastic\Agent\data\tmp\elastic-agent-install135747269\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64: Access is denied.: rename C:\Program Files\Elastic\Agent\data\elastic-agent-af32fe\install\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64 C:\Program Files\Elastic\Agent\data\tmp\elastic-agent-install135747269\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64: Access is denied.
2021-03-09T14:05:33.106-0800    ERROR   log/reporter.go:36      2021-03-09T14:05:33-08:00: type: 'ERROR': sub_type: 'FAILED' message: Application: endpoint-security--7.12.0-SNAPSHOT[15e46460-8123-11eb-b9c4-e16b4b3269c3]: State changed to FAILED: rename C:\Program Files\Elastic\Agent\data\elastic-agent-af32fe\install\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64 C:\Program Files\Elastic\Agent\data\tmp\elastic-agent-install135747269\endpoint-security-7.12.0-SNAPSHOT-windows-x86_64: Access is denied.
2021-03-09T14:05:34.036-0800    INFO    log/reporter.go:40      2021-03-09T14:05:34-08:00: type: 'STATE': sub_type: 'RUNNING' message: Application: [15e46460-8123-11eb-b9c4-e16b4b3269c3]: State changed to RUNNING:
```

**Comment by EricDavisX** [comment]
@ph @andresrc fyi.  I submit it is 'high' impact and if the fix is well understood and not too impactful that we could try to pick it up in 7.12.  Michal is reviewing.

Also, of the 3 'BUG' statements that Dan passed along, I submit we use this ticket only to fix the first, the core 'upgrade' did not happen for endpoint bug.  The other 2 are related feature / bugs about the Health of Agent and relying on assumed success instead of a polling / verification system to manage if Agent managed processes and tasks initiated were successful.  A larger discussion, not sure if we have a ticket for that, though I know we were working generally on better 'health status' as part of Fleet Server work.

I"ll try to reproduce on a Windows template Michal and I both have.

**Comment by EricDavisX** [comment]
I reproduced this with a Win 10 host.  Interestingly, I first tried a win 7 host, and it worked just fine.

Here are some more logs if helpful:
[logs.zip](https://github.com/elastic/beats/files/6120071/logs.zip)

Separately, on an 8.0 system I tried an upgrade from 7.11 to 8.0 and it failed, but Michal mentioned it seemed a different failure, haha.

**Comment by michalpristas** [comment]
i was lucky to reproduce this issue but funny thing is it's not 100% reproducible. 
access denied is not on endpoint directory but on the tmp directory created by our installer used for unpacking zip files then this is renamed and moved to correct path. 

this part was not modified, we added an awaiting layer recently which waits for whole unpack, copy to be finished but i think this was present for longer time and is maybe a timing issue between unpack/sync/rename workflow. ( we sync file by file while unpacking and dir after unpack)

when unpacking we have 1 rename without sync which maybe related. i will raise a PR and test with PR artifacts

**Comment by EricDavisX** [comment]
if we feel confident in the changes we could always merge 8.0 master if we think it won't break anything worse - and then we can test with 'real' 8.0 artifacts from Infra snapshots.

**Comment by EricDavisX** [comment]
We are looking forward to testing in 7.12 BC5.  @dikshachauhan-qasource and team are aware.

**Comment by EricDavisX** [comment]
this is a duplicate to the other known issue:
https://github.com/elastic/beats/issues/24477

it is merged, but the overall snapshot build is failing since March 11 (when this was merged) so it makes sense why it is not picked up there yet.  last passing build:
https://internal-ci.elastic.co/view/Unified%20Release%20-%20snapshot/job/elastic+unified-release+master+distributed-snapshot-multijob-master/307/

closing as such.  please track in the above issue's PR's and if / when needed, watch the unified build process (like above) to see if the build has passed.  let me know any questions, thanks!

**Comment by michalpristas** [comment]
just bare in mind that access denied happens occasionally after rename (which we use a lot during installations). it would be worth to try/test this multiple times to mark it as fixed.

**Comment by michalpristas** [comment]
tried it few times with SNAPSHOT from today and i did not run into the issue, fix should be also present in BC5 as @EricDavisX already mentioned

**Comment by EricDavisX** [comment]
@charlie-pichette in support of testing this, which

**Comment by EricDavisX** [comment]
accidentally closed - reopened it now.  sorry.

**Comment by dikshachauhan-qasource** [comment]
Hi @EricDavisX  cc @michalpristas 

We have validated above fixes in 8.0 snapshot build and found it **fixed**. Build details are as follows:

```
BUILD 41248
COMMIT 716e2f78166346267bae9ae2190c2d35cad8ec01
```

Observation: Endpoint is now also getting upgraded via agent.  
- Agent is healthy.
- No error messages were recorded in logs.
- All binaries [filebeat, metricbeat and endpoint] were in running state.

Please refer below screenshot:
![image](https://user-images.githubusercontent.com/72375798/111454914-32cc4480-873b-11eb-957d-51362aec57f6.png)


Please let us know if anything is missed.

Thanks
QAS

**Comment by dikshachauhan-qasource** [comment]
### `Bug Conversion`

-  01 new testcase is created under the 'Fleet>Upgrade Agent' section:
   - https://elastic.testrail.io/index.php?/cases/view/147373

Thanks!
